src/dtx/dtx_coll.c

```diff
@@ -0,0 +1,373 @@
+
+		goto out;
+	}
+
+	dcpa->dcpa_result = vos_dtx_load_mbs(cont->sc_hdl, &dci->dci_xid, &dcpa->dcpa_oid, &mbs);
+	if (dcpa->dcpa_result == -DER_INPROGRESS && !dtx_cont_opened(cont) &&
+	    opc == DTX_COLL_CHECK) {
+		rc = start_dtx_reindex_ult(cont);
+		if (rc != 0)
+			D_ERROR(DF_UUID": Failed to trigger DTX reindex: "DF_RC"\n",
+				DP_UUID(cont->sc_uuid), DP_RC(rc));
+	}
+
+	if (dcpa->dcpa_result != 0)
+		goto out;
+
+	dcpa->dcpa_result = dtx_coll_prep(dci->dci_po_uuid, dcpa->dcpa_oid, &dci->dci_xid, mbs, -1,
+					  dci->dci_version, cont->sc_pool->spc_map_version,
+					  opc == DTX_COLL_CHECK, false, &dcpa->dcpa_dce);
+	if (dcpa->dcpa_result != 0)
+		D_ERROR("Failed to prepare the bitmap (and hints) for collective DTX "
+			DF_DTI" opc %u: "DF_RC"\n", DP_DTI(&dci->dci_xid), opc,
+			DP_RC(dcpa->dcpa_result));
+
+out:
+	if (cont != NULL)
+		ds_cont_child_put(cont);
+
+	rc = ABT_future_set(dcpa->dcpa_future, NULL);
+	D_ASSERT(rc == ABT_SUCCESS);
+}
+
+int
+dtx_coll_prep(uuid_t po_uuid, daos_unit_oid_t oid, struct dtx_id *xid, struct dtx_memberships *mbs,
+	      uint32_t my_tgtid, uint32_t dtx_ver, uint32_t pm_ver, bool for_check, bool need_hint,
+	      struct dtx_coll_entry **p_dce)
+{
+	struct pl_map		*map = NULL;
+	struct pl_obj_layout	*layout = NULL;
+	struct pool_target	*target;
+	struct dtx_daos_target	*ddt;
+	struct dtx_coll_target	*dct;
+	struct dtx_coll_entry	*dce = NULL;
+	struct daos_obj_md	 md = { 0 };
+	uint32_t		 node_nr;
+	d_rank_t		 my_rank = dss_self_rank();
+	d_rank_t		 max_rank = 0;
+	int			 rc = 0;
+	int			 i;
+	int			 j;
+
+	D_ASSERT(mbs->dm_flags & DMF_COLL_TARGET);
+
+	D_ALLOC_PTR(dce);
+	if (dce == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	dce->dce_xid = *xid;
+	dce->dce_ver = dtx_ver;
+	dce->dce_refs = 1;
+
+	ddt = &mbs->dm_tgts[0];
+	dct = (struct dtx_coll_target *)(ddt + mbs->dm_tgt_cnt);
+	D_ALLOC(dce->dce_bitmap, dct->dct_bitmap_sz);
+	if (dce->dce_bitmap == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	dce->dce_bitmap_sz = dct->dct_bitmap_sz;
+
+	if (!for_check) {
+		memcpy(dce->dce_bitmap, dct->dct_tgts + dct->dct_tgt_nr, dct->dct_bitmap_sz);
+	} else {
+		map = pl_map_find(po_uuid, oid.id_pub);
+		if (map == NULL) {
+			D_ERROR("Failed to find valid placement map in pool "DF_UUID"\n",
+				DP_UUID(po_uuid));
+			D_GOTO(out, rc = -DER_INVAL);
+		}
+
+		for (i = 0, j = 0; i < dct->dct_tgt_nr; i++) {
+			rc = pool_map_find_target(map->pl_poolmap, dct->dct_tgts[i], &target);
+			D_ASSERT(rc == 1);
+
+			/* Skip the targets that reside on other engines. */
+			if (unlikely(target->ta_comp.co_rank != my_rank))
+				continue;
+
+			/* Skip the target that (re-)joined the system after the DTX. */
+			if (target->ta_comp.co_ver > dtx_ver)
+				continue;
+
+			/* Skip non-healthy one. */
+			if (target->ta_comp.co_status != PO_COMP_ST_UP &&
+			    target->ta_comp.co_status != PO_COMP_ST_UPIN &&
+			    target->ta_comp.co_status != PO_COMP_ST_NEW &&
+			    target->ta_comp.co_status != PO_COMP_ST_DRAIN)
```

Comment:
Before the target is fully integrated into the pool, its status will be UP instead of UPIN, right? If yes, the UP target will get RPC. As for "NEW", I am not quite sure. @wangdi1 , any idea for that?